### PR TITLE
tarsnap service: fix multiple simultaneous archives with a single key

### LINKED ISSUE #15948: tarsnap often wants fsck
```
Jun 02 09:50:18 guava tarsnap@-start[14392]: tarsnap: Sequence number mismatch: Run --fsck
Jun 02 09:50:18 guava tarsnap@-start[14392]: tarsnap: Error creating new archive
Jun 02 09:50:18 guava tarsnap@-start[14392]: tarsnap: Error exit delayed from previous errors.
```

I wonder if we should run `fsck` by default in the service? This is very common and blocks my backup, which has poor visibility unless I'm fiddling with systemd.

@thoughtpolice thoughts?

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

Fixes #15948. Notice that this removes option `services.tarsnap.cachedir`, instead adding `services.tarsnap.archives.<name>.cachedir`. The latter is more meaningful now when we know that cache directories correspond to keys.

As a bonus, we now automatically create cache directory on the first run and achieve possibility of starting concurrent archiving jobs via locking.
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by abbradar** [comment]
Other possibly important non-compatibilities is that there are now multiple services named `tarsnap-smth` instead of one generic `tarsnap-${smth}` -- this is needed to substitute cache directory and other values properly.

**Comment by abbradar** [comment]
@domenkozar Do you think it's okay to backport this to 16.09? It removes an option, but the release is technically not there yet ~_^.

**Comment by domenkozar** [comment]
if @thoughtpolice reviews it :)

**Comment by joachifm** [comment]
If this isn't backported, is there a way we can communicate to users that multiple cache dirs on the same key may cause backups to fail?

**Comment by abbradar** [comment]
We can add an assertion -- I can do this if patch won't get through.

**Comment by domenkozar** [comment]
@abbradar yeah let's do that for 16.09 - too late now.

File: nixos/modules/services/backup/tarsnap.nix [hunk 1]
```diff
@@ -262,72 +256,82 @@ in
...
+          NoNewPrivileges = "true";
+          CapabilityBoundingSet = [ "CAP_DAC_READ_SEARCH" ];
+          PermissionsStartOnly = "true";
+        };
+      }) gcfg.archives;
 
     # Note: the timer must be Persistent=true, so that systemd will start it even
     # if e.g. your laptop was asleep while the latest interval occurred.
-    systemd.timers = mapAttrs' (name: cfg: nameValuePair "tarsnap@${name}"
+    systemd.timers = mapAttrs' (name: cfg: nameValuePair "tarsnap-${name}"
```
Comment (domenkozar) [inline nixos/modules/services/backup/tarsnap.nix]: I fail to understand why is this needed?

**Comment by abbradar** [comment]
I've finally found time to add this assertion -- see 444246113a5e57f3ec489baa73bacbe95ee40e00

**Comment by domenkozar** [comment]
@thoughtpolice any comments on this one?

**Comment by abbradar** [comment]
Fixed the conflict.
